src/query_plan/query_planning_traversal.rs

```diff
@@ -87,17 +89,72 @@ struct OpenBranchAndSelections {
     selections: Vec<NormalizedSelection>,
 }
 
-struct BestQueryPlanInfo {
+pub(crate) struct BestQueryPlanInfo {
     /// The fetch dependency graph for this query plan.
-    fetch_dependency_graph: FetchDependencyGraph,
+    pub fetch_dependency_graph: FetchDependencyGraph,
     /// The path tree for the closed branch options chosen for this query plan.
-    path_tree: OpPathTree,
+    pub path_tree: OpPathTree,
     /// The cost of this query plan.
-    cost: QueryPlanCost,
+    pub cost: QueryPlanCost,
 }
 
-impl QueryPlanningTraversal {
-    fn find_best_plan(&mut self) -> Result<Option<&BestQueryPlanInfo>, FederationError> {
+impl BestQueryPlanInfo {
+    // PORT_NOTE: The equivalent of `createEmptyPlan` in the JS codebase.
+    pub fn empty(parameters: &QueryPlanningParameters) -> Self {
+        Self {
+            fetch_dependency_graph: FetchDependencyGraph::new(
+                parameters.supergraph_schema.clone(),
+                parameters.federated_query_graph.clone(),
+                None,
+                0,
+            ),
+            path_tree: OpPathTree::new(parameters.federated_query_graph.clone(), parameters.head),
+            cost: Default::default(),
+        }
+    }
+}
+
+impl<'a> QueryPlanningTraversal<'a> {
+    pub fn new(
+        // TODO(@goto-bus-stop): This probably needs a mutable reference for some of the
+        // yet-unimplemented methods, and storing a mutable ref in `Self` here smells bad.
+        // The ownership of `QueryPlanningParameters` is awkward and should probably be
+        // refactored.
+        parameters: &'a QueryPlanningParameters,
+        _selection_set: NormalizedSelectionSet,
+        has_defers: bool,
+        root_kind: SchemaRootDefinitionKind,
+        cost_processor: FetchDependencyGraphToCostProcessor,
+    ) -> Self {
+        // FIXME(@goto-bus-stop): Is this correct?
+        let is_top_level = parameters.head_must_be_root;
+        Self {
+            parameters,
+            root_kind,
+            has_defers,
+            starting_id_generation: 0,
+            cost_processor,
+            is_top_level,
+            // TODO(@goto-bus-stop): Use `self.resolve_condition_plan()` once it exists
+            condition_resolver: CachingConditionResolver,
+            open_branches: Default::default(),
+            closed_branches: Default::default(),
+            best_plan: None,
+        }
+    }
+
+    // PORT_NOTE: In JS, the traversal is still usable after finding the best plan. Here we consume
+    // the struct so we do not need to return a reference.
+    //
+    // XXX(@goto-bus-stop): Do we still need it? It's easier if we can consume the traversal
```

Comment:
from JS code it looks that after plan is calculated the planning traversal is discarded and is not reused by anything -> I think we can just return the best plan